### PR TITLE
Remove fuzzy

### PR DESCRIPTION
**Pull request by michelbl**
Connected to #437

### CODE REVIEW

File: openfisca_core/legislations.py [hunk 1]
```diff
@@ -305,9 +303,12 @@ def generate_dated_bracket_json(bracket_json, instant_str):
 
 def generate_dated_json_value(values_json, instant_str):
     for value_json in values_json:
-        value_stop_str = value_json.get('stop')
-        if value_json['start'] <= instant_str and (value_stop_str is None or instant_str <= value_stop_str):
-            return value_json['value']
+        value_start_str = value_json.get('start')
```
Comment (fpagnoux) [inline openfisca_core/legislations.py]: These `_str` are useless and make it hard to read this code.

**Comment by michelbl** [comment]
I found a (dirty) way to ckeck the tests with travis : 
- Add the flag `--process-dependency-links` in `.travis.yml` : `- pip install --process-dependency-links --editable .[test]` This flag is deprecated, however no replacement is available yet.
- In `setup.py`, add `dependency_links = ['git+https://github.com/michelbl/openfisca-dummy-country.git@remove-fuzzy#egg=OpenFisca-Dummy-Country-0.2.0'],` The version number (0.2.0) can be anything but it is required.

**Comment by fpagnoux** [comment]
To test:

- For a basic parameter with no end date or placeholder:
    - Check that requesting it before its first start period raises an error
    - Check that its value change at the expected time

- For a basic parameter with a placeholder:
    - Check that requesting after the placeholder date does not raise an error

- For a basic parameter with an end date:
    - Check that requesting after the end date does raise an error

File: openfisca_core/reforms.py [hunk 2]
```diff
@@ -126,58 +126,43 @@ def build_node(node, path_index):
...
-    if not_contained_items and not_contained_items[-1].get('stop') is None \
-            and str(start) > not_contained_items[-1]['start']:
-        last_not_contained_item = not_contained_items[-1].copy()  # Copy to avoid modifying `items` argument.
-        last_not_contained_item['stop'] = str(start.offset(-1, 'day'))
-        not_contained_items[-1] = last_not_contained_item
-    new_items = sorted(not_contained_items + [new_item], key=lambda item: item['start'])
-    return new_items
-
-
-def create_item(start, stop, value):
```
Comment (fpagnoux) [inline openfisca_core/reforms.py]: If you are deprecating and removing this method, you must mention it in the changelog.

File: openfisca_core/reforms.py [hunk 3]
```diff
@@ -126,58 +126,43 @@ def build_node(node, path_index):
...
-    dict_items = []
-    if start is not None:
-        dict_items.append(('start', str(start)))
-    if stop is not None:
-        dict_items.append(('stop', str(stop)))
-    dict_items.append(('value', value))
-    return collections.OrderedDict(dict_items)
-
-
-def split_item_containing_instant(instant, items):
```
Comment (fpagnoux) [inline openfisca_core/reforms.py]: Same here

File: openfisca_core/reforms.py (hunk 2 above)
Comment (fpagnoux) [inline openfisca_core/reforms.py]: @benjello is `create_item` still used to build reforms ?

File: openfisca_core/reforms.py (hunk 3 above)
Comment (fpagnoux) [inline openfisca_core/reforms.py]: @benjello is `split_item_containing_instant ` still used to build reforms ?

**Comment by fpagnoux** [comment]
I rebased the branch on the latest master, you may have to `git reset --hard origin/remove-fuzzy` instead of `git pull` (careful, you can lose local changes on your local branch `remove-fuzzy` if you have not yet pushed them).

File: openfisca_core/reforms.py (hunk 2 above)
Comment (michelbl) [inline openfisca_core/reforms.py]: ok

File: openfisca_core/reforms.py (hunk 3 above)
Comment (michelbl) [inline openfisca_core/reforms.py]: ok

**Comment by michelbl** [comment]
@fpagnoux good point! The exception was not raised because undefined brackets at a given instant where not discared.

This is now fixed. However, the test whether a bracket is defined for a given instant is : https://github.com/openfisca/openfisca-core/pull/488/files#diff-3de96604c5f1d166b9a78364d4584288R357
If, for example, the threshold is correctly ended but not the rate, the bracket will be invalid but not ignored.

I was not able to find a list of rules that must be verified for scales validate_brackets_json_dates, except the code of `validate_brackets_json_dates` https://github.com/openfisca/openfisca-core/blob/master/openfisca_core/legislations.py#L879

I certainly should dig into that code and reimplement it without the fuzzy system. However that is a consequent amount of work. Is it worth it ?

**Comment by fpagnoux** [comment]
The goal will be to replace the current way of writing them by something like [this](https://api-test.openfisca.fr/parameter/impot_revenu.bareme), so let's not spend too much time validating a syntax we want to depreciate.

**Comment by fpagnoux** [comment]
Waiting for the tests of https://github.com/openfisca/openfisca-france/pull/731/files to pass to be convinced this is fully functional.

**Pull request merged**